### DISCUSSION TITLE
What is idiomatic way to signal fx app to stop when server errors

### DISCUSSION THREAD

**Discussion by anivanovic** [uber-go/fx#848] [Q&A] [answered]
As titled my question is what is preferred way to notify fx app to stop if I get error from `server.ListenAndServe()` method? Currently my code registers bunch of Providers and one Invoke which starts up my http server. Example code is like this:

```
fx.Invoke(startServer)

func startServer(lc fx.Lifecycle, router *chi.Mux, cfg *config.Env, app *models.Application) error {
	server := &http.Server{
		Addr:         fmt.Sprintf(":%s", 8080),
		Handler:      router,
	}

	lc.Append(fx.Hook{
		OnStart: func(ctx context.Context) error {
			go func() {
				if err := server.ListenAndServe(); err != nil {
					log.Err(err).Msg("Application server error while running")
					// notify app to stop
					syscall.Kill(os.Getpid(), syscall.SIGINT)
				}
			}()

			log.Info().Msgf("Application ready on port %v", port)
			return nil
		},
		OnStop: func(ctx context.Context) error {
			log.Info().Msg("Application shutting down")
			return multierr.Combine(
				server.Shutdown(ctx),
				app.Close(),
			)
		},
	})

	return nil
}
```

If I get error from `server.ListenAndServe()` I use `syscall.Kill()` to notify fx to stop application and exit. I run app with `fxApp.Run()`. It there better way to do this?

**Comment by anivanovic** [comment]
@sywhang I am tagging you as I see you did most recent changes. Is there recommended way to do this in fx or this is currently the best way?

**Comment by abhinav** [comment, marked as answer]
Hi there! Typically, you can use the [fx.Shutdowner](https://pkg.go.dev/go.uber.org/fx#Shutdowner) type to signal that the app should be shut down. This interface is provided to your container by default so you can just add it as a parameter to your provied/invoked function.

**However**, for the case of `net/http.Server`, note that [per the documentation](https://pkg.go.dev/net/http#Server.ListenAndServe) that `Sever.ListenAndServe` *always* returns a non-nil error—even if you called `Shutdown` or `Close`. So you probably want to ignore ErrServerClosed:

```go
if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed)  {
  shutdowner.Shutdown()
}
```

Hope this helps!

**Reply by anivanovic** [reply]
Thanks on your reply. I would suggest to add this to example in the docs also as I could not figure out this one on my own.
